File: src/main/java/net/shadowfacts/underwaterutilities/UnderwaterUtilities.java
package net.shadowfacts.underwaterutilities;

import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.common.capabilities.CapabilityManager;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
import net.minecraftforge.fml.common.network.NetworkRegistry;
import net.shadowfacts.underwaterutilities.api.item.Goggles;
import net.shadowfacts.underwaterutilities.block.ModBlocks;
import net.shadowfacts.underwaterutilities.event.ClientEventHandler;
import net.shadowfacts.underwaterutilities.gui.UUGUIHandler;
import net.shadowfacts.underwaterutilities.item.ModItems;
import net.shadowfacts.underwaterutilities.recipe.ModRecipes;
import net.shadowfacts.shadowmc.capability.Storage;

/**
 * @author shadowfacts
 */
@Mod(modid = UnderwaterUtilities.modId, name = UnderwaterUtilities.name, version = UnderwaterUtilities.version, dependencies = "required-after:shadowmc;", acceptedMinecraftVersions = "[1.10.2]", guiFactory = "net.shadowfacts.underwaterutilities.gui.UUGUIFactory")
public class UnderwaterUtilities {

//	TODO: set min shadowmc version

	public static final String modId = "underwaterutilities";
	public static final String name = "Underwater Utilities";
	public static final String version = "@VERSION@";

	@Mod.Instance(modId)
	public static UnderwaterUtilities instance;

//	Content
	public static ModBlocks blocks = new ModBlocks();
	public static ModItems items = new ModItems();

	@Mod.EventHandler
	public void preInit(FMLPreInitializationEvent event) {
		UUConfig.init(event.getModConfigurationDirectory());
		UUConfig.load();

		blocks.init();
		items.init();

		NetworkRegistry.INSTANCE.registerGuiHandler(instance, new UUGUIHandler());

		registerCapabilities();

		MinecraftForge.EVENT_BUS.register(new ClientEventHandler());
	}

	@Mod.EventHandler
	public void init(FMLInitializationEvent event) {
		ModRecipes.init();
	}

	private void registerCapabilities() {
		CapabilityManager.INSTANCE.register(Goggles.class, new Storage<>(), Goggles.class);
	}

}

File: src/main/kotlin/net/shadowfacts/underwaterutilities/UnderwaterUtilities.kt
package net.shadowfacts.underwaterutilities

import net.minecraft.block.Block
import net.minecraft.item.Item
import net.minecraftforge.client.event.ModelRegistryEvent
import net.minecraftforge.common.capabilities.CapabilityManager
import net.minecraftforge.event.RegistryEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.network.NetworkRegistry
import net.minecraftforge.fml.common.registry.GameRegistry
import net.shadowfacts.shadowmc.capability.NoOpStorage
import net.shadowfacts.underwaterutilities.api.item.BreathingAid
import net.shadowfacts.underwaterutilities.api.item.Goggles
import net.shadowfacts.underwaterutilities.block.UUBlocks
import net.shadowfacts.underwaterutilities.gui.UUGUIHandler
import net.shadowfacts.underwaterutilities.item.UUItems

/**
 * @author shadowfacts
 */
@Mod(modid = MOD_ID, name = NAME, version = VERSION, dependencies = "required-after:tesla;required-after:shadowmc@[3.4.8,);", guiFactory = "net.shadowfacts.underwaterutilities.gui.UUGUIFactory", modLanguageAdapter = "net.shadowfacts.forgelin.KotlinAdapter", updateJSON = "https://update.shadowfacts.net/underwater-utilities")
object UnderwaterUtilities {

	//	Content
	var blocks = UUBlocks
	var items = UUItems

	@Mod.EventHandler
	fun preInit(event: FMLPreInitializationEvent) {
		UUConfig.init(event.modConfigurationDirectory)
		UUConfig.load()

		NetworkRegistry.INSTANCE.registerGuiHandler(this, UUGUIHandler)

		CapabilityManager.INSTANCE.register(Goggles::class.java, NoOpStorage<Goggles>(), Goggles::class.java)
		CapabilityManager.INSTANCE.register(BreathingAid::class.java, NoOpStorage<BreathingAid>(), BreathingAid::class.java)
	}

	@Mod.EventBusSubscriber
	object RegistrationHandler {

		@JvmStatic
		@SubscribeEvent
		fun registerBlocks(event: RegistryEvent.Register<Block>) {
			event.registry.registerAll(
					blocks.collector,
					blocks.tank
			)

			GameRegistry.registerTileEntity(blocks.collector.tileEntityClass, blocks.collector.name)
			GameRegistry.registerTileEntity(blocks.tank.tileEntityClass, blocks.tank.name)
		}

		@JvmStatic
		@SubscribeEvent
		fun registerItems(event: RegistryEvent.Register<Item>) {
			event.registry.registerAll(
					blocks.collector.createItemBlock(),
					blocks.tank.createItemBlock(),
					items.goggles,
					items.goggleLens,
					items.breather,
					items.tank,
					items.blade,
					items.fan,
					items.snorkelTube,
					items.snorkel
			)
		}

		@JvmStatic
		@SubscribeEvent
		fun registerModels(event: ModelRegistryEvent) {
			blocks.collector.initItemModel()
			blocks.tank.initItemModel()
			items.goggles.initItemModel()
			items.goggleLens.initItemModel()
			items.breather.initItemModel()
			items.tank.initItemModel()
			items.blade.initItemModel()
			items.fan.initItemModel()
			items.snorkelTube.initItemModel()
			items.snorkel.initItemModel()
		}

	}

}
